### PR TITLE
add devcontainer

### PR DESCRIPTION
**Pull request by berrydenhartog**
## Description

added devcontainers. statically setting the version of the the OS (Debian12), python and node for developers. also booting up mongo and redis automaticly. 

NOTE: this PR depends on https://github.com/IMDA-BTG/aiverify/pull/267 to be merged for the redis connection to work

## Motivation and Context

For a more universal standardized development environment on different OS system it would be benificial to start using devcontainers. 

## How to Test

Open you Code Editor and install the devcontainer plugin. 

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [X]  I have tested the changes locally and verified that they work as expected.
- [ ]  I have added or updated the necessary documentation (README, API docs, etc.).
- [x]  I have added appropriate unit tests or functional tests for the changes made.
- [x]  I have followed the project's coding conventions and style guidelines.
- [X]  I have rebased my branch onto the latest commit of the main branch.
- [X]  I have squashed or reorganized my commits into logical units.
- [x]  I have added any necessary dependencies or packages to the project's build configuration.
- [X]  I have performed a self-review of my own code.
- [X]  I have read, understood and agree to the Developer Certificate of Origin below, which this project utilises.

## Screenshots (if applicable)

[If the changes involve visual modifications, include screenshots or GIFs that demonstrate the changes.]

## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]

<details>
  <summary>Developer Certificate of Origin</summary>

 ```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
 ```
</details>

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+.devcontainer
+.github
+.mypy_cache
+deploy-scripts
+examples
+setup-ai-verify
+third-party
```
Comment (anneschuth) [inline .dockerignore:7]: Can we have files end with a (single) EOL?

```suggestion
third-party

```

File: .gitignore [hunk 2]
```diff
@@ -3,4 +3,11 @@ coverage
...
-.tox
\ No newline at end of file
+.tox
+
+# vscode ignores
+.vscode/*
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
```
Comment (anneschuth) [inline .gitignore:13]:
```suggestion
!.vscode/extensions.json

```

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        },
+        {
+            "name": "test-engine-app",
+            "type": "debugpy",
+            "request": "launch",
+            "cwd": "${workspaceFolder}/test-engine-app",
+            "module": "test_engine_app"
+        }
+    ]
+}
```
Comment (anneschuth) [inline .vscode/launch.json:29]:
```suggestion
}

```

File: scripts/devcontainer-post-create.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+find $PROJECT_DIR/stock-plugins/ -mindepth 1 -maxdepth 1 -type d | xargs -I{} basename {} | xargs -I{} unzip $PROJECT_DIR/stock-plugins/{}/dist/*.zip -d $PROJECT_DIR/ai-verify-portal/plugins/{}
+
+# install plugins dependencies
+find $PROJECT_DIR/stock-plugins/ -type f -name 'requirements.txt' -exec pip install -r "{}" \;
+
+
+
+
+
+
```
Comment (anneschuth) [inline scripts/devcontainer-post-create.sh:44]:
```suggestion

```

File: scripts/devcontainer-post-start.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+#npm run dev &
+nodemon app.js &
+
+cd $PROJECT_DIR/test-engine-app
+python -m test_engine_app &
+
+cd $PROJECT_DIR/ai-verify-portal
+npm run dev &
+
+sleep infinity
```
Comment (anneschuth) [inline scripts/devcontainer-post-start.sh:16]:
```suggestion
sleep infinity

```

File: scripts/redis-init/redis.conf [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+# Write logs to stdout
+#logfile /var/log/redis/redis-server.log
+logfile ""
+
+databases 16
+
+# Enable notification for HSET
+notify-keyspace-events Kh
+
+
```
Comment (anneschuth) [inline scripts/redis-init/redis.conf:15]:
```suggestion
```

**Review by anneschuth — COMMENTED** [review]
LGTM AFAIK

File: .dockerignore (hunk 1 above)
Comment (berrydenhartog) [reply .dockerignore]: solved

File: .gitignore (hunk 2 above)
Comment (berrydenhartog) [reply .gitignore]: solved

File: .vscode/launch.json (hunk 3 above)
Comment (berrydenhartog) [reply .vscode/launch.json]: solved

File: scripts/devcontainer-post-create.sh (hunk 4 above)
Comment (berrydenhartog) [reply scripts/devcontainer-post-create.sh]: solved

File: scripts/devcontainer-post-start.sh (hunk 5 above)
Comment (berrydenhartog) [reply scripts/devcontainer-post-start.sh]: solved

File: scripts/redis-init/redis.conf (hunk 6 above)
Comment (berrydenhartog) [reply scripts/redis-init/redis.conf]: solved

File: .devcontainer/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,25 @@
+ARG DEVCONTAINER_VERSION=bullseye
+FROM mcr.microsoft.com/devcontainers/base:${DEVCONTAINER_VERSION}
+
+ENV REDIS_URI=redis://redis:6379
+ENV REDIS_SERVER_HOSTNAME="redis"
+ENV REDIS_URI=redis://redis:6379
```
Comment (uittenbroekrobbert) [inline .devcontainer/Dockerfile:6]: deze regel (6) is een duplicaat van regel 4

**Review by uittenbroekrobbert — COMMENTED** [review]
Looks great, only 1 very small remark, awesome feature!

File: .devcontainer/Dockerfile (hunk 7 above)
Comment (uittenbroekrobbert) [inline .devcontainer/Dockerfile:6]:
```suggestion
```

Comment (berrydenhartog) [reply .devcontainer/Dockerfile:6]: Nice catch. Updated

**Comment by timlrx** [comment]
Some comments trying to get the dev container setup working on my side:

1. @imda-lionelteo could we change [test-engine-app requirements.txt](https://github.com/IMDA-BTG/aiverify/blob/main/test-engine-app/requirements.txt#L17) to `yarl==1.8.1`. Version `1.9.4` is not compatible with the version of `aiopenapi3` used in [test-engine-core-module](https://github.com/IMDA-BTG/aiverify/blob/main/test-engine-core-modules/requirements.txt) and throws an installation error.
2. I tried to run the services using the provided script i.e. `/docker-entrypoint/devcontainer-post-start.sh` but there are some errors with redis: `The middleware stream connection failed to set up properly: The middleware is unreachable. There was a connection error to the redis server: Error 99 connecting to localhost:6379. Cannot assign requested address.`. @berrydenhartog were you able to get it running successfully on your end?

**Comment by berrydenhartog** [comment]
> Some comments trying to get the dev container setup working on my side:
> 
>     1. @imda-lionelteo could we change [test-engine-app requirements.txt](https://github.com/IMDA-BTG/aiverify/blob/main/test-engine-app/requirements.txt#L17) to `yarl==1.8.1`. Version `1.9.4` is not compatible with the version of `aiopenapi3` used in [test-engine-core-module](https://github.com/IMDA-BTG/aiverify/blob/main/test-engine-core-modules/requirements.txt) and throws an installation error.
> 
>     2. I tried to run the services using the provided script i.e. `/docker-entrypoint/devcontainer-post-start.sh` but there are some errors with redis: `The middleware stream connection failed to set up properly: The middleware is unreachable. There was a connection error to the redis server: Error 99 connecting to localhost:6379. Cannot assign requested address.`. @berrydenhartog were you able to get it running successfully on your end?

@timlrx. This PR depends on https://github.com/IMDA-BTG/aiverify/pull/267.  to bypass using that PR you could change localhost to redis on this line https://github.com/IMDA-BTG/aiverify/blob/ff4958e93c6467cc17ff7dff9699542d8bcfaa31/test-engine-app/test_engine_app/config/environment_variables.py#L17

**Comment by anneschuth** [comment]
Or we could base this PR on #267

**Comment by timlrx** [comment]
Thanks @berrydenhartog, the patch works well for me. Also tested the changes with #267 and verified that it works.

@kimeetok let's replace the current dev setup with this and update the docs to recommend devcontainer as the preferred approach?

**Review by imda-benedictlee — APPROVED** [review]
Approved

**Comment by berrydenhartog** [comment]
@imda-benedictlee Seems like the CI/CD wont run because the build can not handle forked repos:
```
[28/65] RUN git clone https://github.com/imda-btg/aiverify.git --branch=feature/devcontainer aiverify:
0.225 Cloning into 'aiverify'...
0.335 fatal: Remote branch feature/devcontainer not found in upstream origin
```

**Comment by imda-benedictlee** [comment]
Hi @berrydenharto, PR #282 has been merged to Main. You can proceed to pull from Main for the new Docker Integration Test Changes. Thanks.

**Comment by berrydenhartog** [comment]
> Hi @berrydenharto, PR #282 has been merged to Main. You can proceed to pull from Main for the new Docker Integration Test Changes. Thanks.

Thanks. Rebased on main

**Review by imda-kelvinkok — APPROVED** [review]

**Comment by imda-benedictlee** [comment]
Hi @berrydenhartog, my apologies, can you help me to remove line 191 in integration-test.yml as it causes the workflow to look for the secret token. `token: ${{ secrets.CHECKOUT_TOKEN }}`. Thanks!

**Comment by imda-benedictlee** [comment]
Hi @berrydenhartog, I realised that the secret token requirement was still in Integration Test Workflow in Main. I have updated the Integration Test Workflow in Main. Apologies, require you to pull from main again.

**Comment by berrydenhartog** [comment]
@imda-benedictlee. I rebased the PR

**Review by imda-benedictlee — APPROVED** [review]
Approved

**Comment by imda-benedictlee** [comment]
Pipeline Testing for Dev Script Passed
[Pipeline Testing For Dev Script.zip](https://github.com/IMDA-BTG/aiverify/files/15074938/Pipeline.Testing.For.Dev.Script.zip)

**Comment by imda-benedictlee** [comment]
SCA Workflow still requires some refactoring. Can ignore SCA Failure.

**Comment by anneschuth** [comment]
Amazing! Thanks for working with us! 🎉

**Comment by imda-benedictlee** [comment]
Hi @anneschuth, thanks to you and the team for contributing and for the help as well 😀

**Pull request merged**